neptune_optuna/impl/__init__.py

```diff
@@ -144,342 +175,484 @@ def __init__(self,
 
     def _log_best_trials(self, study):
-        self.run['best'] = _stringify_keys(_log_best_trials(study))
+        _log_best_trials(self.run, study, self._namespaces)
 
     def _log_study_details(self, study, trial):
         if trial._trial_id == 0:
             _log_study_details(self.run, study)
 
     def _log_plots(self, study, trial):
         if self._should_log_plots(study, trial):
-            _log_plots(self.run, study,
-                       visualization_backend=self._visualization_backend,
-                       log_plot_contour=self._log_plot_contour,
-                       log_plot_edf=self._log_plot_edf,
-                       log_plot_parallel_coordinate=self._log_plot_parallel_coordinate,
-                       log_plot_param_importances=self._log_plot_param_importances,
-                       log_plot_pareto_front=self._log_plot_pareto_front,
-                       log_plot_slice=self._log_plot_slice,
-                       log_plot_optimization_history=self._log_plot_optimization_history,
-                       log_plot_intermediate_values=self._log_plot_intermediate_values,
-                       )
+            _log_plots(
+                self.run,
+                study,
+                visualization_backend=self._visualization_backend,
+                log_plot_contour=self._log_plot_contour,
+                log_plot_edf=self._log_plot_edf,
+                log_plot_parallel_coordinate=self._log_plot_parallel_coordinate,
+                log_plot_param_importances=self._log_plot_param_importances,
+                log_plot_pareto_front=self._log_plot_pareto_front,
+                log_plot_slice=self._log_plot_slice,
+                log_plot_optimization_history=self._log_plot_optimization_history,
+                log_plot_intermediate_values=self._log_plot_intermediate_values,
+                namespaces=self._namespaces,
+            )
 
     def _log_study(self, study, trial):
         if self._should_log_study(trial):
             _log_study(self.run, study)
 
-    def _should_log_plots(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial):
+    def _should_log_plots(
+        self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial
+    ):
         if not len(study.get_trials(states=(optuna.trial.TrialState.COMPLETE,))):
             return False
-        elif self._plots_update_freq == 'never':
+        elif self._plots_update_freq == "never":
             return False
         else:
             if trial._trial_id % self._plots_update_freq == 0:
                 return True
         return False
 
     def _should_log_study(self, trial: optuna.trial.FrozenTrial):
-        if self._study_update_freq == 'never':
+        if self._study_update_freq == "never":
             return False
-        else:
-            if trial._trial_id % self._study_update_freq == 0:
-                return True
+        if trial._trial_id % self._study_update_freq == 0:
```

Comment:
The same here. I'm a little bit afraid of using `_trial_id`. Remember to test it :wink: